text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+   name: 'author',
+   type: 'model' (the default type for the models collection)
+
+src/ui/components/titleize (with `export let helper = Ember.Helper.helper(function() { })`)
+=> collection: 'ui/components',
+   name: 'titleize',
+   type: 'helper'
+
+src/ui/components/show-title (with `export let template = Ember.HTMLBars.template(COMPILED)`)
+=> collection: 'ui/components',
+   name: 'show-title',
+   type: 'template'
+```
+
+### Main Modules
+
+Every project must have a "main" module, named `src/main.js`, that
+serves as an entry point into the project.
+
+The main module must export an `Application`, `Engine`, or (new) `Addon` class.
+This class must define a `modulePrefix`, which must match the node package name
+for the project.
+
+The main module also declares other properties that help the Ember resolver
+understand relationships between projects. For instance, the main module can
+declare which modules in an addon are available to a consuming app's resolver.
+
+The main module of an addon can also declare a `rootName`, which is used by the
+resolver to lookup main modules. Initially, the `rootName` will be a read-only
+property that equals the `modulePrefix` with any `ember-` and  `ember-cli-`
+prefixes stripped (e.g. `ember-power-select` becomes `power-select`). It's
+possible that we may allow overrides / aliases in the future.
+
+Modules that appear alongside `main.js` in `src` are also considered `main`
+modules for their respective `type`. For instance, `src/router.js` is registered
+with a `name` of `main` and a `type` of `router`.
+
+### Module Collections
+
+Top-level namespaces within `src` serve to group modules into
+type-aware "collections".
+
+The following rules apply to module collections and types:
+
+1. Each collection can contain one or more types. The types allowed
+   in a particular collection MUST be explicitly declared.
+2. Each type MAY exist in any number of collections.
+3. Each type MUST have only one "definitive collection", which is the
+   collection the resolver will use for resolutions if a module can't be found
+   in the local (i.e. originating) collection.
+4. Each collection MAY have a single "default type". If a module does not
+   indicate its type through its file name, then its default export should
+   align with the default type for its collection.
+5. Each collection can allow "private collections" to be defined at a namespace.
+   Private collections are localized additions to a top-level collection,
+   available only from the namespace at which they're defined.
+6. Top-level collections may be grouped for organization purposes. No
+   resolvable modules must be placed in a group directory.
+7. A collection can appear only once in a project (i.e. it can not be
+   contained in multiple group directories, or in a group as well as at the
+   top-level).
+
+The following collections and allowed types (rules 1 & 2) are proposed:
+
+* `components` - COMPONENT, HELPER, template
```

Comment:
I feel similarly here. When I was picking up ember, reasoning about why I should write something as a component or a helper wasn't very clear. I defaulted to making everything a component and then spent some time porting several to helpers after running into glimmers too-many-components rendering issues. This change feels like it may make the distinction even less clear.